<!DOCTYPE html>
{% extends "base.html" %}
{% block title %}
    <title>Help | EpiML@ShiLab</title>
{% endblock %}
{% block styles %}
    {{ super() }}
    <style>
        ul {
            font-size: large;
        }

        ol {
            font-size: large;
            counter-reset: item
        }

        li {
            display: block
        }

        li.mlutilLevel:before {
            content: counters(item, ".") " ";
            counter-increment: item
        }
    </style>

    <style type="text/css">
        .tg {
            border-collapse: collapse;
            border-spacing: 0;
        }

        .tg td {
            font-family: Arial, sans-serif;
            font-size: 14px;
            padding: 10px 5px;
            border-style: solid;
            border-width: 1px;
            overflow: hidden;
            word-break: normal;
            border-color: black;
        }

        .tg th {
            font-family: Arial, sans-serif;
            font-size: 14px;
            font-weight: normal;
            padding: 10px 5px;
            border-style: solid;
            border-width: 1px;
            overflow: hidden;
            word-break: normal;
            border-color: black;
        }

        .tg .tg-s6z2 {
            text-align: center
        }
    </style>
    <link rel="stylesheet" type="text/css"
          href="{{ url_for('static',filename='css/back_to_top.css') }}">
{% endblock %}
{% block scripts %}
    {{ super() }}
    <script src='{{ url_for('static', filename='js/offset_height_for_help_nav.js') }}'></script>
    <script src='{{ url_for('static', filename='js/back_to_top.js') }}'></script>

    <!--for math equation -->
    <script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML'
            async></script>
{% endblock %}

{% block content %}

    <!-- for back2top -->
    <a id="back2Top" title="Back to top" href="#">&#10148;</a>

    <h2>Help</h2>
    <hr>
    <div>
        <h3>Table of Contents</h3>
        <ol>
            <li class="mlutilLevel"><a href="#introduction">Introduction</a></li>
            <li class="mlutilLevel"><a href="#getting_started">Getting Started</a>
                <ol>
                    <li class="mlutilLevel"><a href="#prepare">Prepare your data</a></li>
                    <li class="mlutilLevel"><a href="#upload"> Upload data</a></li>
                    <li class="mlutilLevel"><a href="#submit"> Check job status and retrieve results</a></li>
                </ol>
            </li>
            <li class="mlutilLevel"><a href="#result_visualization">Result and visualization</a>
                <ol>
                    <li class="mlutilLevel"><a href="#summary"> Summary</a></li>
                    <li class="mlutilLevel"><a href="#table"> Interactive table</a></li>
                    <li class="mlutilLevel"><a href="#visualization"> Visualization</a></li>
                </ol>
            </li>
            <li class="mlutilLevel"><a href="#customize_code">Customize Code</a>
                <ol>
                    <li class="mlutilLevel"><a href="#install"> Install Docker</a></li>
                    <li class="mlutilLevel"><a href="#generate"> Generate job notebook</a></li>
                    <li class="mlutilLevel"><a href="#launch"> Launch a Docker Container</a></li>
                    <li class="mlutilLevel"><a href="#customize"> Customize your Notebook</a></li>
                </ol>
            </li>
            <li class="mlutilLevel"><a href="#analysis_methods">Analysis Methods</a>
                <ol>
                    <!--<li class="mlutilLevel"><a href="#preprocessing"> Preprocessing</a></li>-->
                    <li class="mlutilLevel"><a href="#pipeline"> Workflow</a></li>
                    <li class="mlutilLevel"><a href="#method"> Methods</a></li>
                </ol>
            </li>
            <li class="mlutilLevel"><a href="#FAQ">FAQ</a></li>
            <li class="mlutilLevel"><a href="#reference">Reference</a></li>
        </ol>
    </div>
    <hr>
    <div>
        <main>
            <section class="anchor" id="introduction">
                <h3>1. Introduction</h3>
                <p>Epistasis describes a phenomenon that different genetic variants affect phenotypes jointly instead of
                    independently. Epistasis among different variants contributes to a large proportion of the hidden in
                    heritability of complex traits [1-3], and plays an important role in gene expression and regulation
                    [4-6]. However, genomic data is typically high-dimensional, where the number of features (genetic
                    variants) is usually much larger than sample size. The large number of genetic variants that present
                    in high dimensional genomic data make it extremely challenging for epistasis analysis since the
                    searching space for epistasis grows exponentially with the number of genetic variants. Penalized
                    regression based machine learning methods work well with analyzing high-dimensional and large-scale
                    genomic data in that they can use penalized terms to impose sparsity to simultaneously enable
                    scalable feature selection and dimension reduction to avoid overfitting. Such methods including
                    classical Least Absolute Shrinkage and Selection Operator (LASSO) [7], newly proposed spike-and-slab
                    LASSO generalized linear models (ssLASSO GLMs) [8, 9] and Empirical Bayesian Elastic Net (EBEN) [10,
                    11] are proved to well scalable the large dimension genomic data in recent studies. Nonetheless,
                    these methods usually include multiple hyperparameters that need to be carefully trained
                    for optimization on a particular genomic dataset. The intrinsic complexity of
                    optimizing these models limits their wide adoption and makes it hard to reproduce results.
                </p>
                <!--<p> Epistasis was originally a concept from genetics in which it refers to the phenomenon that the
                    effect of one gene (locus) is dependent on the presence of one or more 'modifier genes'. As the
                    study of genetics developed, and with the advent of molecular biology, epistasis starts to be
                    studied in relation to Quantitative Trait Loci (QTL) and polygenic inheritance. Increasingly
                    sophisticated computational and evolutionary biology models aim to describe the effects of epistasis
                    on a genome-wide scale and the consequences of this for evolution. Thus, identification of epistatic
                    pairs has large research significance.</p>-->
                <p>
                    <!--We proposed an open machine learning web service for epistasis analysis, call EpiML, which can
                    detect main effect and epistasis effect by using several statistical and machine learning methods,
                    and
                    produce many interactive visualizations. Especially for some case studies like SNP and miRNA data,
                    external resources are embedded for epistasis analysis. EpiML -->

                    To support reproducible epistasis analysis on high dimensional genomic datasets, we have developed
                    an integrated and user-friendly web server for epistasis analysis using classical and emerging
                    penalized machine learning methods (EpiML), which has many advantages: 1) it is easy to
                    use. The users only need to upload their data, select a method and submit their job. Our server will
                    handle all the processing in background. 2) It will provide many friendly interactive
                    visualizations, like circle network, adjacent matrix and force-directed network. 3) For some case
                    study, multiple external resources are integrated in results visualization for analysis. 4) User can
                    customize their analysis. For each job, its corresponding Jupyter notebook is automatically
                    generated. All notebooks can be downloaded, modified and rerun, allowing users to fully customize
                    their epistasis analysis on local computers.
                </p>
                <!--
                <p> Epistasis analysis is used to determine if genes with related mutant phenotypes act
                    in the same or different pathways, and, if in the same pathway, to place them in a
                    linear order relative to one another based on the step in the pathway controlled by
                    that gene. In other words, one uses epistasis analysis to construct an order-offunction
                    map that reflects the sequence of events in a pathway controlled by several
                    genes. [Ref. Genetic Techniques for Biological Research]
                </p>
                -->
                <p> The schematic diagram of EpiML is following:</p>
                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img width="80%" src="{{ url_for('static',filename='img/help/epiml.png') }}">
                        <figcaption style="width: 80%; text-align: justify"><strong>Figure 1. The workflow of
                            EpiML.</strong> <strong>a</strong>. Our epistasis analysis server allows users describe
                            their job, upload data and select a machine learning method for analysis. <strong>b</strong>.
                            We provide interactive
                            tools for visualizing and downloading results. <strong>c</strong>. Customize analysis. All
                            automatically generated Jupyter notebooks and pre-configured Docker containers can be
                            downloaded, modified and rerun, allowing users to fully customize their analysis code on
                            local computers. <!--<strong>d</strong>. Generate bioRxiv report that provides detailed
                            information
                            about the models.-->
                        </figcaption>
                    </figure>
                </div>

            </section>
            <hr>
            <section class="anchor" id="getting_started">
                <h3>2. Getting Started</h3>
                <p> This is a step-by-step tutorial how to use EpiML to find out main and epistasis effects in your
                    data.
                </p>

                <h4 class="anchor" id="prepare">2.1. Prepare your data</h4>
                <p>Input files include two parts: feature matrix <strong>X</strong>, and label matrix <strong>Y</strong>.
                    Both of them are tab-delimited tables with row and column names. In generally, the row names and
                    column names could be any string, but no repeat. However, for gene and microRNA case studies, the
                    column names of feature matrix are very important, because they will be used for parsing the gene
                    position for genotype and searching miRBase for microRNA.
                </p>
                <p>The column names of genotype are combined with SNP id, chromosome id, position, reference allele, and
                    alternative allele, separated by ‘_’. An example for genotype data format is shown in following:
                </p>
                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img width="90%"
                             src="{{ url_for('static',filename='img/help/genotype_feature_matrix_X.png') }}">
                        <figcaption style="width: 90%; text-align: center"><strong>Figure 2. An example of genotype
                            data format.</strong></figcaption>
                    </figure>
                </div>
                <p>The column names of microRNA are microRNA ID in miRBase. An example for microRNA data format is shown
                    in following:
                </p>
                <div align="center">
                    <figure>
                        <img width="90%"
                             src="{{ url_for('static',filename='img/help/microRNA_feature_matrix_X.png') }}">
                        <figcaption style="width: 90%; text-align: center"><strong>Figure 3. An example of microRNA
                            data format.</strong></figcaption>
                    </figure>
                </div>


                <p>The format of label matrix Y for all kind of data are same. An example of label matrix Y is shown in
                    following:
                </p>
                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img height="200px" src="{{ url_for('static',filename='img/help/label_matrix_Y.png') }}">
                        <figcaption style="width: 90%; text-align: center"><strong>Figure 4. An example of label matrix
                            format.</strong></figcaption>
                    </figure>
                </div>

                <p>Important notes:</p>
                <ul>
                    <li>Only files with .txt or .csv extensions are accepted.</li>
                    <li>File size is limited to no more than 5MB.</li>
                </ul>
                <!--
                    <p>
                    <pre>plink --ped mystudy_chr1.ped --map mystudy_chr1.map --recode vcf --out mystudy_chr1</pre>
                    <code>test </code>
                    </p>
                -->
                <h4 class="anchor" id="upload">2.2. Upload data</h4>
                <p>Some basic information is required before you upload data.</p>
                <ul>
                    <li><strong>Job name</strong> is used to identify jobs when you retrieve results.
                    </li>
                    <li><strong>Email</strong> is used to accept notification and retrieve results. It is optional.
                    <li><strong>Job Category</strong> refers to what kind of job. For Gene and microRNA category, links
                        to meta-data and database will be integrated in result visualization.
                    </li>
                    <li><strong>Data Type of Feature Matrix</strong> is used to preprocess feature matrix. For
                        continuous data, it will be
                        quantile normalized.
                    </li>
                    <li><strong>Description</strong> is used to help user annotate job details, such as purpose, type of
                        data, and parameters. It is optional.
                    </li>
                    <li><strong>Upload</strong> the prepared files. Upload the feature matrix file and label matrix file
                        separately.
                    </li>
                    <li><strong>Select Method</strong>.Three popular methods are built in web server. Refer to <a
                            href="#machine_learning_methods">Machine Learning Methods</a> section to learn more details.
                        The cross validation is specifically set as 5.
                    </li>
                </ul>
                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img width="90%" src="{{ url_for('static',filename='img/help/upload_data.png') }}">
                        <figcaption style="width: 90%; text-align: center"><strong>Figure 5. A screenshot of upload
                            form.</strong></figcaption>
                    </figure>
                </div>
                <h4 class="anchor" id="submit">2.3. Check job status and retrieve results</h4>
                <p>Once your job is submitted, it will be added into job queue waiting for execution in background. The
                    browser will jump to a processing page where you can check your job status, as following figure
                    shown. The results will be shown on your screen when it is finished. Because all jobs are running in
                    background, you can bookmark and close this page. You also can reopen this page anytime to check job
                    status. Meanwhile, if you provide an Email address, you will receive an Email notification with this
                    page link.
                </p>
                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img width="90%" src="{{ url_for('static',filename='img/help/processing.png') }}">
                        <figcaption style="width: 90%; text-align: center"><strong>Figure 6. A screenshot of processing
                            page.</strong></figcaption>
                    </figure>
                </div>
            </section>
            <hr>
            <section class="anchor" id="result_visualization">
                <h3>3. Result and visualization</h3>
                <h4 class="anchor" id="summary">3.1. Summary</h4>
                <p>Once the submitted job is finished, you can download results
                    through corresponding links, or directly check your results using following interactive tables and
                    visualizations.
                </p>
                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img width="90%" src="{{ url_for('static',filename='img/help/summary.png') }}">
                        <figcaption style="width: 90%; text-align: justify"><strong>Figure 7. The summary form of
                            finished job.</strong> It contains job description, method, running time, input and output
                            files, and jupyter notebook. User can download files through corresponding link.
                        </figcaption>
                    </figure>
                </div>
                <h4 class="anchor" id="table">3.2. Interactive table</h4>
                <p>EpiML can detect significant main effects and epistasis effects. These detected results will be list
                    in two interactive tables correspondingly, where you can rank whole table by anyone column and
                    search all contents to find out your interesting features.
                </p>

                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img width="90%" src="{{ url_for('static',filename='img/help/result_table.png') }}">
                        <figcaption style="width: 90%; text-align: center"><strong>Figure 8. These detected results
                            will be list in two interactive tables, by which user can easily find their interested
                            information.</strong>
                        </figcaption>
                    </figure>
                </div>

                <h4 class="anchor" id="visualization">3.3. Visualization</h4>
                <p> We provides three kind of interactive visualizations, including circle network, adjacency matrix and
                    force-directed graph. When you move mouse over one node, more detailed information and linked nodes
                    will be highlighted.
                </p>
                <h4>Circle Network</h4>
                <p>One node in circle network represents a feature in detected effects. All nodes are
                    arranged as a circle. Two nodes that have epistatic interaction are linked by an undirected edge.
                    Circle network has two main advantages comparing to table:
                    1) It is easy to find out those features that have multiple epistatic effects with other features;
                    2) It is easy to find indirect relationship among features.</p>
                <p> You can move mouse over one node to find its interaction nodes in red. When you
                    click one node, the node and its associated nodes will be highlighted in green. </p>
                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img width="90%" src="{{ url_for('static',filename='img/help/circle_network.png') }}">
                        <figcaption style="width: 90%; text-align: justify"><strong>Figure 9. An example of circle
                            network based on detected yeast epistasis effects.</strong> Move mouse over one node to find
                            its epistatic nodes in red. When you click one node, the node and its associated nodes will
                            be highlighted in green.
                        </figcaption>
                    </figure>
                </div>

                <h4>Adjacency Matrix</h4>
                <p>The row and column names in adjacency matrix both are the features in detected epistatic effects.
                    Each
                    cell \((i, j)\) represents an edge from vertex \(i\) to vertex \(j\). Matrix cell depicts
                    coefficient value in
                    epistasis effect table. The darker cells indicate strong interaction. Adjacency matrix has two
                    advantages comparing to circle network:
                    1) Adjacency matrix can give a whole picture of epistatic relationship, especially for large number
                    of epistatic features; 2) Matrix cell value can depict the strength of epistatic relationship.
                </p>

                <p> You can move mouse over one node to check its epistatic feature and coefficient value.</p>
                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img width="90%" src="{{ url_for('static',filename='img/help/adjacency_matrix.png') }}">
                        <figcaption style="width: 90%; text-align: justify"><strong>Figure 10. An example of adjacency
                            matrix based on detected yeast epistatic effects.</strong> Move mouse over one node to check
                            its epistatic feature and coefficient value.
                        </figcaption>
                    </figure>
                </div>


                <h4>Force-directed Graph</h4>
                <p>Force-directed graph is a more flexible way to represent epistatic relationship. One node in graph
                    represents a feature in epistasis effect table. Two nodes that have epistatic relationship are
                    linked by an edge. All kinds of information can be
                    embedded in force-directed graph, like coefficient value and external resources.</p>
                <p> You can move mouse over and drag one node to check its linked epistatic features. Click one node to
                    show more information.</p>
                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img width="90%" src="{{ url_for('static',filename='img/help/forced-direct_graph.png') }}">
                        <figcaption style="width: 90%; text-align: justify"><strong>Figure 10. An example of
                            force-directed graph based on detected yeast epistasis effects.</strong> Move mouse over and
                            drag one node to check its linked epistatic features. Click one node to show more
                            information.
                        </figcaption>
                    </figure>
                </div>

            </section>
            <hr>
            <section class="anchor" id="customize_code">
                <h3>4. Customize Code</h3>
                <p>Users can customize analysis code based on released Jupyter notebooks and pre-configured Docker
                    containers. For each job, its corresponding Jupyter notebook is automatically generated. All
                    notebooks can be downloaded, modified and rerun based on pre-configured Docker container which
                    contains a fully functional Jupyter server with all the Python and R libraries necessary to run the
                    notebooks.</p>

                <h4 class="anchor" id="install">4.1. Install Docker</h4>
                <p>Docker is a Container-as-a-Service (CaaS) platform that enables a secure software supply chain and
                    deploys diverse applications for high availability across disparate infrastructure.
                    Download, install and run Docker on your computer from the official
                    <a href="https://www.docker.com/products/docker-engine" target="_blank">Docker website</a> depending
                    on your OS.</p>

                <h4 class="anchor" id="generate">4.2. Generate job notebook</h4>
                <p>Submit your job on EpiML web server. Once the job is done, you can download it from
                    job summary table. If your data is too large, you can submit the demo data following <a
                            href="#getting_started">Get Started</a> section. </p>

                <h4 class="anchor" id="launch">4.3. Launch a Docker Container</h4>
                <ol>
                    <li style="display: list-item"> Open a new terminal window.</li>
                    <li style="display: list-item">Change work directory into that includes your job notebook. Strongly
                        recommend to create a new
                        directory, and copy notebook into the new directory.
                    </li>
                    <li style="display: list-item">Launch the Docker container by entering the following command
                        <pre>docker run -it --rm -p 8888:8888 -v "$PWD":/home/jovyan/work workhardplayharder/epiml:latest</pre>
                        This command pulls the <code>workhardplayharder/epiml</code> image tagged <code>latest</code>
                        from Docker Hub if it is not already present on the local host. It then starts a container
                        running a Jupyter Notebook server, and exposes the server on host port 8888. The server logs
                        appear in the terminal.
                    </li>
                </ol>
                <h4 class="anchor" id="customize">4.4. Customize your Notebook</h4>
                <p>Copy/paste <code>http://&lt;hostname&gt;:8888/?token=&lt;token&gt;</code> in a browser
                    loads the Jupyter Notebook dashboard page, where hostname is the name of the computer running
                    docker and token is the secret token printed in the console. Click into <code>work</code> directory,
                    and find the jupyter notebook. Customize and save your notebook in
                    Jupyter. Once you are done, stop the Docker container by pressing <code>Ctrl+C</code> on the
                    previously opened Terminal window.</p>
            </section>
            <hr>
            <section class="anchor" id="analysis_methods">
                <h3>5. Analysis Methods</h3>
                <p>We build-in three penalized regression based machine learning methods for epistatic analysis. In
                    general, you only need to
                    upload your data, select one machine learning method, and then submit your job. You don't need to
                    care about the algorithm detail. Our server will handle all the procedure in backend. For your
                    better understanding the workflow and difference among build-in methods, the build-in methods are
                    briefly introduced.
                </p>
                <p>The model for identifying main and epistasis effect is shown as below:</p>
                <p> $$y= \mu + X \beta_{main} + (X_i \odot X_j) \beta_{epis} + \epsilon$$</p>
                <p> $$i \neq j; i,j \in [1, p] $$</p>
                <p>where \(y\) is the label, \(\mu\) is mean, \(X\) denotes the feature
                    matrix with dimension \((n, p)\), \(X_i, X_j\) denotes the \(i^{th}\) and \(j^{th}\) features in
                    the matrix
                    \(X\). \(\odot\) denotes the symbol for elementwise multiplication of vectors. \(\beta_{main}\)
                    denotes the
                    main effects, \(\beta_{epis}\) denotes the epistatic effects of the \(i^{th}\) and \(j^{th}\)
                    features,
                    and \(\epsilon\) is the residual error which follows the normal distribution \(N(0,\sigma^2)\).
                </p>

                <!--<h4 class="anchor" id="preprocessing">5.1. Preprocessing</h4>-->

                <h4 class="anchor" id="pipeline">5.1. Workflow</h4>
                <p>Penalized regression based machine learning methods work well for high-dimensional data. However,
                    they are designed for solving linear regression problem. We develop a workflow composed of multiple
                    steps for main and epistasis detection, as shown in Figure 11.
                    <!--First, we identify the main effects using a model with only
                    main effect on the phenotype. Second, a corrected phenotype is calculated by removing the
                    significant main effect from the original phenotype. Third, we select features with epistatic effect
                    on the corrected phenotype. Finally, we run the full model on the previously selected main and
                    epistatic features.-->
                </p>
                <ol>
                    <li style="display: list-item"><strong>Identify candidate main effects from feature matrix.</strong>
                        Run penalized regression based machine
                        learning methods on feature matrix, by which some significant effects can be selected as
                        candidate main effects.
                    </li>
                    <li style="display: list-item"><strong>Calculate the phenotype of epistatic effect.</strong> The
                        phenotype of epistatic effect can be
                        calculated by removing those detected main effects from the original phenotype.
                    </li>
                    <li style="display: list-item"><strong>Generate epistasis matrix.</strong> Generate epistasis matrix
                        by pairwise features with element-wise multiplication.
                    </li>
                    <li style="display: list-item"><strong>Identify candidate epistatic effects from epistasis
                        matrix.</strong> Run penalized regression based machine
                        learning methods on epistasis matrix, by which some significant effects can be selected as
                        candidate epistatic effects.
                    </li>
                    <li style="display: list-item"><strong>Identify significant main and epistatic effects.</strong>
                        Rerun penalized regression based machine
                        learning methods on the previously selected candidate main and epistatic effects to identify the
                        final significant main and epistatic effects.
                    </li>
                </ol>


                <p></p>
                <div align="center">
                    <figure style="margin-bottom: 10px">
                        <img width="45%" src="{{ url_for('static',filename='img/help/workflow.png') }}">
                        <figcaption width="90%" style="text-align: justify"><strong>Figure 11. The overall workflow for
                            epistasis detection.</strong> <strong>Step 1:</strong> Identify candidate main effects from
                            feature matrix <strong>Step 2:</strong>
                            Calculate the phenotype of epistatic effect. <strong>Step 3:</strong> Generate epistasis
                            matrix. <strong>Step 4:</strong> Identify candidate
                            epistatic effects from epistasis matrix. <strong>Step 5:</strong> Run the full model to
                            identify significant main and epistatic effects.
                        </figcaption>
                    </figure>
                </div>

                <h4 class="anchor" id="method">5.2. Methods</h4>
                <p>Three popular penalized regression based machine learning methods are used to in our web server to
                    build the above workflow,
                    including Least Absolute Shrinkage and Selection Operator (LASSO) [7], Spike-and-Slab LASSO
                    generalized linear models (ssLASSO GLMs) [8, 9], and Empirical Bayesian Elastic Net (EBEN) [10, 11].
                </p>

                <!--<p>Lasso is based on sparse shrinkage solution where only a limited number of parameters with non-zero
                    coefficients are identified to be associated with the phenotype. EBEN and ssLasso GLMs are Bayesian
                    based methods that impose priors on the parameters.
                </p>-->

                <h4>Least Absolute Shrinkage and Selection Operator (LASSO)</h4>
                <p>LASSO is one of the most commonly used method which introduces a \(L_1\) regularization term that
                    adds a
                    penalty term equivalent to absolute value of coefficients on all parameters. LASSO will randomly
                    select one feature if a set of features are closely correlated. There is one hyperparameter
                    \(\lambda\) to adjust the regularization degree which means it would induce more coefficients are
                    shinked to zero if increasing \(\lambda\). Cross validation strategy is used to choose
                    the \(\lambda\) to optimize the Lasso model [7].
                </p>

                <h4>Spike-and-Slab LASSO (ssLASSO)</h4>
                <p>ssLASSO GLMs imposes different shrinkage priors, spike-and-slab mixture double-exponential prior
                    distribution, on related and unrelated parameters which can induce optimal model solutions. For the
                    ssLASSO GLMs, \(y\) is represented by a link function that
                    \(\eta = E(y_i|X_i)\). ssLasso GLMs combines the EM algorithm and cyclic coordinate descent
                    algorithm to fastly fit high-dimension data. There are two hyperparameters (\(s_0\) and \(s_1\)) in
                    ssLASSO algorithm. The \(s_0\) is a
                    small value and serves as a “spike” scale for modelling irrelevant coefficients and inducing strong
                    shrinkage, while \(s_1\) is chosen to be relatively large value and serves to be “slab” scale for
                    modelling related coefficients and inducing no or weak shrinkage. The grid-search based method can
                    be applied to select the optimal combination of two hyperparameters. In an alternative way, it’s
                    reported that \(s_1\) has little influence on results comparing to \(s_0\), then users can use
                    generalized linear model (GLM) to estimate the data scale to be as \(s_0\) [8, 9].
                </p>

                <h4>Empirical Bayesian Elastic Net (EBEN)</h4>
                <p>EBEN is a Bayesian based method which use a fast coordinate ascent method and doesn’t reply on MCMC
                    procedure. Hence, EBEN can scalable the over-saturated genetic model to estimate the parameters in
                    the model efficiently. In addition, EBEN can conduct the hypothesis test such as \(t\)-test using
                    the covariance matrix and the coefficients which is different from elastic net method [5]. There are
                    two hyperparameters, such as \(\lambda_1\) and \(\lambda_2\), are the components of a generalized
                    Gamma distribution which imposes on the unknown variance of each coefficients. Cross-validation
                    strategy is used to tune these two hyperparameters to optimize the EBEN model [10, 11].
                </p>
            </section>
            <hr>
            <section class="anchor" id="FAQ">
                <h3>6. FAQ</h3>
                <ol>
                    <li style="display: list-item">How long does it take to get results?</li>
                    <p>The processing time depends on the scale of upload data and selected method. In generally, the
                        LASSO and ssLASSO are very fast. EBEN takes long time, especially number of folds is large. For
                        hundreds of samples and features, LASSO and ssLASSO can be finished in several minutes.
                        EBEN may need an hour. </p>
                    <li style="display: list-item">Why can't I upload files?</li>
                    <p>First, please check your file extension. Only files with .txt or .csv extensions are accepted.
                        Second, file size is limited to no more than 5MB due to time consuming. If your data is too
                        large, please follow customize code section, and run the pipeline on local host.</p>
                    <li style="display: list-item">Which browser is recommended?</li>
                    <p>EpiML is tested to work on the following browsers. We recommend using <a
                            href="https://www.google.com/chrome/" target="_blank">Chrome</a> or <a
                            href="https://www.apple.com/safari/" target="_blank">Safari</a>. If you are already using
                        one of the above browsers try updating it. If it still doesn't work let us know.</p>

                    <table class="tg">
                        <tr>
                            <th class="tg-s6z2">OS</th>
                            <th class="tg-s6z2">Version</th>
                            <th class="tg-s6z2">Chrome</th>
                            <th class="tg-s6z2">Firefox</th>
                            <th class="tg-s6z2">Microsoft Edge</th>
                            <th class="tg-s6z2">Safari</th>
                        </tr>
                        <tr>
                            <td class="tg-s6z2">Linux</td>
                            <td class="tg-s6z2">Ubuntu 18.04</td>
                            <td class="tg-s6z2">70.0</td>
                            <td class="tg-s6z2">63.0</td>
                            <td class="tg-s6z2">NA</td>
                            <td class="tg-s6z2">NA</td>
                        </tr>
                        <tr>
                            <td class="tg-s6z2">MacOS</td>
                            <td class="tg-s6z2">Mojave</td>
                            <td class="tg-s6z2">70.0</td>
                            <td class="tg-s6z2">63.0</td>
                            <td class="tg-s6z2">NA</td>
                            <td class="tg-s6z2">12.0</td>
                        </tr>
                        <tr>
                            <td class="tg-s6z2">Windows</td>
                            <td class="tg-s6z2">10</td>
                            <td class="tg-s6z2">70.0</td>
                            <td class="tg-s6z2">63.0</td>
                            <td class="tg-s6z2">44.0</td>
                            <td class="tg-s6z2">NA</td>
                        </tr>
                    </table>
                </ol>

                <p>For any more question about this service please email to <a href="mailto: x.shi@uncc.edu"> Xinghua
                    Shi</a> or <a href="mailto: jchen67@uncc.edu"> Junjie Chen</a>
                </p>
            </section>
            <section class="anchor" id="reference">
                <h3>7. Reference</h3>

                <ol>
                    <li style="display: list-item">
                        O Carlborg and C. S. Haley. Epistasis: Too often neglected in complex trait studies? Nature
                        Reviews Genetics, 5:618625, 2004.
                    </li>
                    <li style="display: list-item">
                        Greg Gibson. Hints of hidden heritability in GWAS. Nat Genet, 42(7):558–560, July 2010.
                    </li>
                    <li style="display: list-item">
                        Zuk O, Hechter E, Sunyaev SR, and Lander ES. The mystery of missing heritability: Genetic
                        interactions create phantom heritability. Proc Natl Acad Sci U S A., 109(4):1193–8, 2008.
                    </li>
                    <li style="display: list-item">
                        G. Gibson. Epistasis and pleiotropy as natural properties of transcriptional regulation. Theor.
                        Popul. Biol., 49:5889, 1996.
                    </li>
                    <li style="display: list-item">
                        G.W. et al. Carter. Prediction of phenotype and gene expression for combinations of mutations.
                        Mol. Syst. Biol., 3:96, 2007.
                    </li>
                    <li style="display: list-item">
                        J. et al. Gertz. Epistasis in a quantitative trait captured by a molecular model of
                        transcription factor interactions. Theor. Popul. Biol., 77:15, 2010.
                    </li>
                    <li style="display: list-item">
                        Tibshirani R. Regression shrinkage and selection via the lasso. Journal of the Royal Statistical
                        Society. Series B (Methodological). 1996; 1:267-88.
                    </li>
                    <li style="display: list-item">
                        Tang Z., Shen Y., Zhang X., Yi N. The spike-and-slab lasso generalized linear models for
                        prediction and associated genes detection. Genetics. 2017; 205(1): 77-88.
                    </li>
                    <li style="display: list-item">
                        Tang Z., Shen Y., Li Y., Zhang X., Wen J., Qian C.A., Zhuang W., Shi X., Yi N. Group
                        spike-and-slab lasso generalized linear models for disease prediction and associated genes
                        detection by incorporating pathway information. Bioinformatics. 2017; 34(6): 901-10.
                    </li>
                    <li style="display: list-item">
                        Huang A., Xu S., Cai X. Empirical Bayesian elastic net for multiple quantitative trait locus
                        mapping. Heredity. 2015; 114(1): 107.
                    </li>
                    <li style="display: list-item">
                        Zou H., Hastie T. Regularization and variable selection via the elastic net. Journal of the
                        Royal Statistical Society: Series B (Statistical Methodology). 2005; 67(2): 301-20.
                    </li>

                </ol>

            </section>


        </main>
    </div>
{% endblock %}